plotly/figure_factory/_facet_grid.py

```diff
@@ -0,0 +1,1073 @@
+             'x': x_pos,
+             'xanchor': 'center',
+             'xref': 'paper',
+             'y': y_pos,
+             'yanchor': 'middle',
+             'yref': 'paper'}
+    return annot
+
+
+def _add_shapes_to_fig(fig, annot_rect_color, flipped_rows=False,
+                       flipped_cols=False):
+    fig['layout']['shapes'] = []
+    for key in fig['layout'].keys():
+        if 'axis' in key and fig['layout'][key]['domain'] != [0.0, 1.0]:
+            if 'xaxis' in key:
+                if flipped_cols:
+                    fig['layout']['shapes'].append(
+                        {'fillcolor': annot_rect_color,
+                       'layer': 'below',
+                       'line': {'color': annot_rect_color, 'width': 1},
+                       'type': 'rect',
+                       'x0': fig['layout'][key]['domain'][0],
+                       'x1': fig['layout'][key]['domain'][1],
+                       'xref': 'paper',
+                       'y0': 1.005,
+                       'y1': 1.05 + 0.5,
+                       'yref': 'paper'}
+                    )
+                else:
+                    fig['layout']['shapes'].append(
+                        {'fillcolor': annot_rect_color,
+                       'layer': 'below',
+                       'line': {'color': annot_rect_color, 'width': 1},
+                       'type': 'rect',
+                       'x0': fig['layout'][key]['domain'][0],
+                       'x1': fig['layout'][key]['domain'][1],
+                       'xref': 'paper',
+                       'y0': 1.005,
+                       'y1': 1.05,
+                       'yref': 'paper'}
+                    )
+            elif 'yaxis' in key:
+                if flipped_rows:
+                    fig['layout']['shapes'].append(
+                        {'fillcolor': annot_rect_color,
+                         'layer': 'below',
+                         'line': {'color': annot_rect_color, 'width': 1},
+                         'type': 'rect',
+                         'x0': 1.005,
+                         'x1': 1.05 + 1,
+                         'xref': 'paper',
+                         'y0': fig['layout'][key]['domain'][0],
+                         'y1': fig['layout'][key]['domain'][1],
+                         'yref': 'paper'}
+                    )
+                else:
+                    fig['layout']['shapes'].append(
+                        {'fillcolor': annot_rect_color,
+                         'layer': 'below',
+                         'line': {'color': annot_rect_color, 'width': 1},
+                         'type': 'rect',
+                         'x0': 1.005,
+                         'x1': 1.05,
+                         'xref': 'paper',
+                         'y0': fig['layout'][key]['domain'][0],
+                         'y1': fig['layout'][key]['domain'][1],
+                         'yref': 'paper'}
+                    )
```

Comment:
This could be drier. It's hard to immediately see which attributes are different for each of these conditonal statements. Something like this might be better:
```
shape = {
    'fillcolor': annot_rect_color,
    'layer': 'below',
    'line': {'color': annot_color}, 
     # etc - the rest of the shared attributes
}
if 'xaxis' in key::
    shape['x0'] = fig['layout'][key]['domain'][0]
    shape['x1'] = fig['layout'][key]['domain'][0]
    shape['y0'] = 1.005
    if flipped_cols:
        shape['y1'] = 1.05 + 0.5
    else:
        shape['y1'] = 1.05
elif 'yaxis' in key:
    shape['x0'] = 1.005
    shape['y0'] = fig['layout'][key]['domain'][0]
    shape['y1'] = fig['layout'][key]['domain'][0]
    if flipped_rows:
        shape['x1'] = 1.05 + 1
     else:
         shape['x1'] = 1.05
```